core/arxiv/submission/services/plaintext/plaintext.py

```diff
@@ -0,0 +1,279 @@
+"""
+Provides integration with the plaintext extraction service.
+
+This integration is focused on usage patterns required by the submission
+system. Specifically:
+
+1. Must be able to request an extraction for a compiled submission.
+2. Must be able to poll whether the extraction has completed.
+3. Must be able to retrieve the raw binary content from when the extraction
+   has finished successfully.
+4. Encounter an informative exception if something goes wrong.
+
+This represents only a subset of the functionality provided by the plaintext
+service itself.
+"""
+
+from typing import Tuple, List, Any, Union, NamedTuple, Optional, Dict
+from urllib.parse import urlparse, urlunparse, urlencode
+from enum import Enum
+from math import exp, log
+import json
+from functools import wraps
+
+from arxiv import status
+from arxiv.taxonomy import Category
+from arxiv.base.globals import get_application_config, get_application_global
+
+import requests
+from requests.packages.urllib3.util.retry import Retry
+
+
+class RequestFailed(IOError):
+    """The plain text extraction service returned an unexpected status code."""
+
+    def __init__(self, msg: str, data: dict = {}) -> None:
+        """Attach (optional) data to the exception."""
+        self.data = data
+        super(RequestFailed, self).__init__(msg)
+
+
+class DoesNotExist(RequestFailed):
+    """The requested resource does not exist."""
+
+
+class ExtractionInProgress(RequestFailed):
+    """An extraction is already in progress."""
+
+
+class RequestUnauthorized(RequestFailed):
```

Comment:
Why are these HTTP 4xx codes essentially replicated here instead of shared with something like `arxiv.base`? What about a `plaintext.RequestUnauthorized` necessitates more code?

Reply:
The goal here is to communicate with the caller (inside the service) about error conditions when using the classifier service integration module. It's true that in this case there is a pretty close mapping between HTTP status codes and exception classes, but that need not be the case. E.g. consider https://github.com/arXiv/arxiv-submission-core/pull/48/files/b3c02004979c20e0ce2a1eac9b95a6e4ef043e38#diff-2e1e1f027b0fdbf696bbfb5b990e0668R61. While strictly speaking you could reduce the overall number of lines of code in the project by moving some of these exceptions to base, that comes at the cost of having a split-brain situation in which some of this module's exceptions are defined in one place, and the remainder defined here. Keeping the exceptions local makes the module a bit more self-contained at a fairly low cost (a few lines of code).